x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/dataframe/transforms/SourceConfig.java

```diff
@@ -0,0 +1,120 @@
+
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.optionalConstructorArg;
+
+
+public class SourceConfig implements Writeable, ToXContentObject {
+
+    public static final ParseField QUERY = new ParseField("query");
+    public static final ParseField INDEX = new ParseField("index");
+
+    public static final ConstructingObjectParser<SourceConfig, Void> STRICT_PARSER = createParser(false);
+    public static final ConstructingObjectParser<SourceConfig, Void> LENIENT_PARSER = createParser(true);
+
+    private static ConstructingObjectParser<SourceConfig, Void> createParser(boolean lenient) {
+        ConstructingObjectParser<SourceConfig, Void> parser = new ConstructingObjectParser<>("data_frame_config_source",
+            lenient,
+            args -> {
+                @SuppressWarnings("unchecked")
+                String[] index = ((List<String>)args[0]).toArray(new String[0]);
+                // default handling: if the user does not specify a query, we default to match_all
+                QueryConfig queryConfig = args[1] == null ? QueryConfig.defaultQueryConfig() : (QueryConfig) args[1];
+                return new SourceConfig(index, queryConfig);
+            });
+        parser.declareStringArray(constructorArg(), INDEX);
+        parser.declareObject(optionalConstructorArg(), (p, c) -> QueryConfig.fromXContent(p, lenient), QUERY);
+        return parser;
+    }
+
+    private final String[] index;
+    private final QueryConfig queryConfig;
+
+    public SourceConfig(String[] index, QueryConfig queryConfig) {
+        ExceptionsHelper.requireNonNull(index, INDEX.getPreferredName());
+        if (index.length == 0) {
+            throw new IllegalArgumentException("must specify at least one index");
+        }
+        if (Arrays.stream(index).anyMatch(i -> i == null || i.isEmpty())) {
+            throw new IllegalArgumentException("all indices need to be non-null and non-empty");
+        }
+        this.index = index;
+        this.queryConfig = ExceptionsHelper.requireNonNull(queryConfig, QUERY.getPreferredName());
+    }
+
+    public SourceConfig(final StreamInput in) throws IOException {
+        index = in.readStringArray();
+        queryConfig = in.readOptionalWriteable(QueryConfig::new);
```

Comment:
It used to be optional prior to this change - see https://github.com/elastic/elasticsearch/pull/40337/files#diff-1e334358634cddafc1877781e52d09acL139

It's certainly wrong to have optional here yet call `requireNonNull` in the constructor.

However, I think there's an important usability issue tied up with this.  Now we have more deeply nested objects we're making it harder for a developer who wants `match_all` if we force non-null objects for the defaults at deeply nested levels.  One option would be to have a constructor for `SourceConfig` that just takes one argument and sets the default query.  Then the HLRC could also provide that constructor but in the HLRC case it would leave `queryConfig` as `null` to pick up the server-side default.

Reply:
@droberts195 there are definitely usability help for the HLRC I will be doing in my next commit. I think we should probably have a `SourceConfig(String... index)`, and possibly make it variadic in the HLRC in the other ctr as well.